### ISSUE TITLE
Using .star file as input

### ISSUE THREAD

**Issue by aamelo** [BradyAJohnston/MolecularNodes#157]
Hi Brady,

Would it be possible to add a function to use Relion's .star as an input for molecular nodes?
.star files have XYZ coordinates and Euler angles for a single structure. 

Best,
Arthur

**Comment by BradyAJohnston** [comment]
Are you able to provide an example file and links to documentation that might explain the file type? I haven't any experience with this file type so we'll need some more details to be able to support it

**Comment by aamelo** [comment]
Sure! I'm attaching one example of a STAR file. The documentation about Relion and the ".star" file convention can be found in this link (https://relion.readthedocs.io/en/release-3.1/Reference/Conventions.html). 

Based on a PDB file, I'm guessing you might only need to use the columns with coordinates (rlnCoordinateX, rlnCoordinateY, rlnCoordinateX), orientations (rlnAngleRot, rlnAngleTilt, rlnAnglePsi) and shifts (rlnOriginXAngst, rlnOriginYAngst, rlnOriginZAngst). 

Also, Alister Burt (https://github.com/alisterburt) has a Python implementation for the STAR format called starfile (https://github.com/teamtomo/starfile).  

[starfile_example.star.txt](https://github.com/BradyAJohnston/MolecularNodes/files/10712426/starfile_example.star.txt)

**Comment by BradyAJohnston** [comment]
Thanks for the additional info. What would be the final result, from importing this kind of file? You would like a series of points / atoms corresponding to each line of the file? Sorry I'm still unsure what result is desired as this seems to be more a type of file that is a series of images / tomograms? I'm sure @alisterburt would happily contribute his expertise to it as well

**Comment by aamelo** [comment]
The final result would be an image of an object, can be an imported map as .stl or a PDB structure, plotted onto each point in the correct orientation. 

I have done something using Chimera (attached) where this is a 3D representation of my sample in the ice. ![example_image](https://user-images.githubusercontent.com/59894900/218240404-0852b9dd-a5e4-4a37-ad80-f4eb81ff8c37.png)

My sample only contains one protein, so the final result here is an image with repeating units of the same map/object in different orientations. I was wondering if that would be possible to do using a PDB, almost like generating a biological assembly from the PDB, but without any symmetry information, only the information in the STAR file.

In the future I was thinking on using Blender for animations of a tomogram of a cell, using a STAR file and multiple PDBs of different proteins. This it would be a great way to visualize and create figures/animations directly using the experimental data.

**Comment by BradyAJohnston** [comment]
ah ok I think I am understanding what you are wanting to achieve. 

Using the star file as a source of translations & rotations to then instance a PDB (imported via MN) should be very doable.

I'll see what Alister might have to say about parsing the files, but we could likely utilise some of the code from starfile to parse it and then create the points with attributes that are usable inside of Geometry Nodes.

**Comment by alisterburt** [comment]
Hi @BradyAJohnston @aamelo - thanks for tagging me in

Here are two scripts for getting the positions and orientations from RELION STAR files as numpy arrays using starfile/scipy - I'm not sure if/how you parametrise 3D rotations in MolecularNodes but this should get us going

gist links: [RELION 3.0](https://gist.github.com/alisterburt/83496cae0cb852fcd149902b53a8eefc) / [RELION 3.1](https://gist.github.com/alisterburt/fe8eb89c4bee793c7a1cda63bdc623e7)

Instancing the PDB would be super cool - can you easily instance meshes too? There will be cases where people have a reconstruction but no molecular model, deriving a mesh from the volume is ~trivial :)

If you point me at the right parts of the codebase I'd love to help make this happen!

**Comment by alisterburt** [comment]
also, cool looking data @aamelo !!

**Comment by BradyAJohnston** [comment]
You can instance any kind of object, whether it be mesh, volume or otherwise so I can imagine this working quite well.

Looking at those gists that you put together, that basically achieves everything we'd need to get the data into Geometry Nodes / Molecular Nodes. All I need is the XYZ positions and the euler rotations and they can be instanced inside of GN. I might open up a PR with some initial boilerplate and get some feedback on the implementation.

As an aside also @alisterburt, I don't have any experience with the volume files that are produced from EM. These should also be able to be imported into Blender in theory (Blender has great support for open VDB as a volume file) but if the volume files can be read in as a 3D numpy array with densities at each point, then I can also use that as a method for importing volumes straight into Blender and building the meshes (with control over density cutoff etc) inside of Blender. Has been something that quite a few people have asked for, but I haven't had any experience with them so I haven't looked into it yet.

**Comment by alisterburt** [comment]
> You can instance any kind of object, whether it be mesh, volume or otherwise so I can imagine this working quite well.

😍 

> All I need is the XYZ positions and the euler rotations and they can be instanced inside of GN.

Sweet - looking at the [Blender docs on Eulers](https://docs.blender.org/manual/en/latest/glossary/index.html#term-Euler) it looks like their convention is XYZ extrinsic (equivalent to ZYX intrinsic). You can construct Euler angles in this convention from the Euler angles in a RELION STAR file as follows

```python
relion_euler_angles = df[['rlnAngleRot', 'rlnAngleTilt', 'rlnAnglePsi']].to_numpy()
blender_euler_angles = R.from_euler(
    seq='ZYZ', angles=euler_angles, degrees=True
).inv().as_euler(seq='xyz', degrees=True)
```

The scipy rotation routines can be a little slow when the sets of euler angles get huge, if that becomes an issue I have a package [eulerangles](https://github.com/alisterburt/eulerangles) which is ~10x faster.

>  if the volume files can be read in as a 3D numpy array with densities at each point, then I can also use that as a method for importing volumes straight into Blender and building the meshes (with control over density cutoff etc) inside of Blender. Has been something that quite a few people have asked for, but I haven't had any experience with them so I haven't looked into it yet.

Super nice - files produced by ~all pieces of software will be in the MRC2014 format (`.mrc`). These can be read into 3D numpy arrays of shape `(nz, ny, nx)` with [mrcfile](https://github.com/ccpem/mrcfile) as follows

```python
import mrcfile

volume = mrcfile.read('/path/to/file.mrc')  # (nz, ny, nx)
```

If you need information about the pixel size of the map you can get it from the header like this

```python
import mrcfile

with mrcfile.open('/path/to/file.mrc', header_only=True, permissive=True) as mrc:
    voxel_size = (mrc.header.nz, mrc.header.ny, mrc.header.nx)
```

Note that the axis order of the arrays is zyx rather than the xyz commonly used for coordinates. I'm not sure of Blenders conventions for mapping volumetric data into its world space, just keep this in mind.

Drop me a note if you need a hand or run into any issues - would love to see this supported and very happy to help make it happen! 🙂

**Comment by alisterburt** [comment]
n.b. if you want to avoid the scipy dependency for whatever reason *eulerangles* is pure numpy

**Comment by BradyAJohnston** [comment]
My numpy (and python in general) skills are still pretty rubbish. If I wanted to take the grid points of an example map (255, 256, 256) and convert it to a numpy array of dimensions `(n, 4)` where `n = 255 * 256 * 256`, what would be the best way of doing this? The components of each row would be `(x, y, z, density)` where the `x, y, z` are the grid points position in arbitrary space (-128 to 127 for example, as this can be scaled later).



I've done some digging around and found that getting access directly to the volume objects inside of Blender currently isn't possible (https://devtalk.blender.org/t/getting-openvdb-grids-in-render-engine-2-83/12270/14?) but I can create my own grid of points and assign density to them the traditional way.

**Comment by BradyAJohnston** [comment]
Okay looks like I have an importer for the starfile, actually super straight forward (thanks to @alisterburt's code). 

Might have PR to review shortly. 

<img width="1792" alt="image" src="https://user-images.githubusercontent.com/36021261/218425748-3d88d717-8212-47fa-a889-e910177083e8.png">

The importing of the volumes is a whole other story. Might open up a new issue about it (harder than initially anticipated).

**Comment by alisterburt** [comment]
Love to see it!!

**Comment by aamelo** [comment]
That looks awesome! Thanks a lot @BradyAJohnston and @alisterburt for implementing it. 

In this case, could I assign any type of object created in blender to the coordinates?

**Comment by BradyAJohnston** [comment]
Yep currently I am instancing a cone, but that can be replaced by any type of objected inside of Blender.

**Comment by aamelo** [comment]
Can we already download it to test?

**Comment by alisterburt** [comment]
@aamelo you would have to install the branch in #161 from source

**Comment by aamelo** [comment]
@alisterburt were you able to install it? I got the #161 branch and installed it in Blender, but nothing shows up in the Add-on panel.

**Comment by alisterburt** [comment]
@aamelo I'm ashamed to admit I have never actually opened molecular nodes - my blender experience is limited to the donut tutorial 😂

**Comment by aamelo** [comment]
@alisterburt That is such a classic tutorial, it was my start on blender as well. @BradyAJohnston tutorial on the lipid bilayer is really good. Let's wait for him then.

**Comment by BradyAJohnston** [comment]
@aamelo to install the file itself, blender will need a `.zip` folder. Inside the repo, if you just compress the `MolecularNodes` folder to a `.zip` then use that to install that should work for Blender.

Otherwise you can use the VSCode plugin to build & install development versions of an addon for Blender, which I highly recommend. (https://github.com/JacquesLucke/blender_vscode)

**Comment by BradyAJohnston** [comment]
addressed in #196
